src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunSingleAssemblyCompilationModuleGroup.cs

```diff
@@ -84,161 +70,12 @@ public sealed override bool ContainsMethodBody(MethodDesc method, bool unboxingS
-                type.IsFunctionPointer ||
-                type.IsCanonicalDefinitionType(CanonicalFormKind.Any))
-            {
-                return true;
-            }
-            var defType = (MetadataType)type;
-            if (!ContainsType(defType))
-            {
-                if (!defType.IsValueType)
-                {
-                    // Eventually, we may respect the non-versionable attribute for reference types too. For now, we are going
-                    // to play is safe and ignore it.
-                    return false;
-                }
-
-                // Valuetypes with non-versionable attribute are candidates for fixed layout. Reject the rest.
-                if (!defType.HasCustomAttribute("System.Runtime.Versioning", "NonVersionableAttribute"))
-                {
-                    return false;
-                }
-            }
-            if (!defType.IsValueType && !ContainsTypeLayout(defType.BaseType))
-            {
-                return false;
-            }
-            foreach (FieldDesc field in defType.GetFields())
-            {
-                if (field.IsStatic)
-                    continue;
-
-                TypeDesc fieldType = field.FieldType;
-
-                if (fieldType.IsValueType && 
-                    !ContainsTypeLayout(fieldType))
-                {
-                    return false;
-                }
-            }
-
-            return true;
-        }
-
-        public override bool VersionsWithModule(ModuleDesc module)
-        {
-            return _versionBubbleModuleSet.Contains(module);
-        }
-
-        public override bool VersionsWithType(TypeDesc typeDesc)
-        {
-            return typeDesc.GetTypeDefinition() is EcmaType ecmaType &&
-                _versionBubbleModuleSet.Contains(ecmaType.EcmaModule);
-        }
-
-        public override bool VersionsWithMethodBody(MethodDesc method)
-        {
-            return VersionsWithType(method.OwningType);
-        }
-
-        public override bool CanInline(MethodDesc callerMethod, MethodDesc calleeMethod)
-        {
-            // Allow inlining if the caller is within the current version bubble
-            // (because otherwise we may not be able to encode its tokens)
-            // and if the callee is either in the same version bubble or is marked as non-versionable.
-            bool canInline = VersionsWithMethodBody(callerMethod) &&
-                (VersionsWithMethodBody(calleeMethod) ||
-                    calleeMethod.HasCustomAttribute("System.Runtime.Versioning", "NonVersionableAttribute"));
-
-            return canInline;
+            return (ContainsType(method.OwningType) && VersionsWithMethodBody(method)) || CompileVersionBubbleGenericsIntoCurrentModule(method);
         }
 
-        public override bool GeneratesPInvoke(MethodDesc method)
+        public sealed override void ApplyProfilerGuidedCompilationRestriction(ProfileDataManager profileGuidedCompileRestriction)
```

Comment:
I would rather see this as a constructor argument because immutable classes tend to be better in multithreaded codebases - or can you at least make this fail if `_profileGuidedCompileRestriction` was already set?

Reply:
I'll make it fail. The reason it isn't a constructor argument, is that the profiler guided infrastructure is going to need to use the CompilationModuleGroup, and it gets ugly fast. However, making the function throw if set is completely a great idea.

Reply:
I'll also make it fail if we try to run down the code that uses this thing without setting it up. It would be bad to use a null profieldata manager and then a real one.